#+TITLE: A*PA: A* Pairwise Aligner
#+PROPERTY: header-args :eval no-export :exports results

A global pairwise sequence aligner using A*.

Co-authored by [[https://github.com/pesho-ivanov][@pesho-ivanov]] and [[https://github.com/RagnarGrootKoerkamp][@RagnarGrootKoerkamp]].

- Usage ::
  ~cargo run --release -- -i <input>.{seq,txt,fasta} -o output.csv~

  This globally aligns sequence pairs from the input and writes a csv with lines of
  ~{cost},{score}~. *You will need Rust nightly!*

  If you run into any kind of problem, please (/please/ ðŸ¥º) make an issue or
  reach out on twitter or matrix (@curious_coding:matrix.org).
- Citations ::
  Please cite the Bioinformatics paper. The bioRxiv version includes the
  supplement and has slightly nicer formatting.
  - BioRxiv preprint:

    *Ragnar Groot Koerkamp*, *Pesho Ivanov*.
    "Exact global alignment using A* with chaining seed heuristic and match pruning".
    bioRxiv (2024). [[https://doi.org/10.1101/2022.09.19.508631][10.1101/2022.09.19.508631]]
  - OUP Bioinformatics paper:

    *Ragnar Groot Koerkamp*, *Pesho Ivanov*.
    "Exact global alignment using a* with chaining seed heuristic and match pruning".
    Bioinformatics (2024). [[https://doi.org/10.1093/bioinformatics/btae032][10.1093/bioinformatics/btae032]]
- Links ::
  - Twitter: [[https://mobile.twitter.com/curious_coding][@curious_coding]], [[https://mobile.twitter.com/peshotrie][@peshotrie]],
  - Blog: [[https://research.curiouscoding.nl]]
- Disclaimer ::
  A*PA can be slow and/or go out of memory when the error rate is high ($\gg
  10\%$) or when long indels are present. You could try
  lower/higher $k$, and more/less pruning, but otherwise use Edlib or BiWFA.

** Example

Here's an A*PA alignment of two sequences of length $500$ with $30\%$ error rate.

[[file:imgs/readme/layers.gif]]

* Usage

** Command line application

- [[https://rustup.rs/][Install rustup]], enable ~nightly~ (~rustup install nightly~ ~rustup default nightly~), and clone this repo.
- ~cargo run --release -- -n 100000~ should already work and run on some randomly generated sequences.
- If you get build errors related to ~sdl2~ and/or missing ~ttf~ files, disable
  the visualizer by passing ~--no-default-features~ to ~cargo~: ~cargo run
  --release --no-default-features -- <flags>~. Alternatively, install ~sdl2~ using e.g. ~apt-get install libsdl2-ttf-dev~.

#+begin_src shell :exports both :results verbatim
cargo run --release -- -h
#+end_src

#+RESULTS:
#+begin_example
Globally align pairs of sequences using A*PA

Usage: pa-bin [OPTIONS] <--input <INPUT>|--length <LENGTH>>

Options:
  -i, --input <INPUT>    A .seq, .txt, or Fasta file with sequence pairs to align
  -o, --output <OUTPUT>  Write a .csv of `{cost},{cigar}` lines
      --no-dt            Do not use the diagonal-transition optimization
  -s, --silent...        Print less stats. Pass twice for summary line only
  -h, --help             Print help (see more with '--help')

Heuristic:
  -H, --heuristic <HEURISTIC>  [default: gcsh] [possible values: none, zero, gap, sh, csh, gcsh,
                               gap-cost, affine]
  -r <r>                       Seed potential [default: 2]
  -k <k>                       Seed length [default: 15]
      --prune <PRUNE>          [default: start] [possible values: none, start, end, both]

Generated input:
  -n, --length <LENGTH>          Target length of each generated sequence [default: 1000]
  -e, --error-rate <ERROR_RATE>  Error rate between sequences [default: 0.05]

Visualizer:
  -v, --visualize <WHEN>  Interactive visualizer. See --help for more [default: none] [possible
                          values: none, first, last, all, layers]
#+end_example

*** Examples
To align all consecutive pairs in a file:
#+begin_src
cargo run --release -- -i <path/to/file.{fasta,fa,txt,seq}> -o cigars.csv
#+end_src
To run on $100$ random sequences of length $10^5$ with error rate $5\%$:
#+begin_src
cargo run --release -- --cnt 100 -n 100000 -e 0.05
#+end_src
Show a video of a small alignment:
#+begin_src sh
cargo run --release -- -n 100 -e 0.10 -v all --style detailed
#+end_src
Save a large alignment to disk:
#+begin_src sh
cargo run --release -- -i <input> --draw Layers --save-last --save-path alignment --style large
#+end_src

*** Unpublished features
- Pass  ~--max-matches <num>~ to use variable length seeds with at most ~<num>~
  matches per seed.  ~--kmin <kmin>~, ~--kmax <kmax>~ are sometimes needed to
  constrain seed lengths.
- Pass ~--skip-prune <N>~ to skip pruning every ~N~'th match that would
  otherwise be pruned. This can speed up pruning when there are a lot of matches.

** Rust API
The ~astarpa~ crate is the [[file:astarpa/src/lib.rs][main entrypoint]]. See the docs there.
Use ~astarpa::astarpa(a, b)~ for alignment with default settings or
~astarpa::astarpa_gcsh(a,b,r,k,end_pruning)~ to use GCSH+DT with custom parameters.

Examples of more advanced usages using the ~AstarPa~ aligner object can be found at [[file:pa-bin/examples/][pa-bin/examples]].

** C API
The ~astarpa-c~ [[file:astarpa-c/astarpa.h][crate]] contains simple C-bindings for the
~astarpa::{astarpa,astarpa_gcsh}~ functions and an [[file:astarpa-c/example.c][example]] with [[file:astarpa-c/makefile][makefile]]. More should not be needed for
simple usage.

* Visualization

Currently only A*PA can be visualized using the binary. Reimplementations of
Needleman-Wunsch, band-doubling (Edlib), and diagonal-transition (WFA, BiWFA)
are available in the ~pa-base-algos~ crate and can be called from rust for now;
see the [[file:pa-bin/examples/paper-figures/intro.rs][examples]].

Sample videos corresponding to figure 1 of the paper are below. Due to different
visualization strategies (per layer, per cell) timings are not comparable.

|----------------------------------------------------------------------+----------------------------------------------------------------------------|
| Dijkstra [[file:imgs/readme/2_dijkstra.gif]]                             | Ukkonen's exponential search (Edlib) [[file:imgs/readme/1_ukkonen.gif]]        |
| Diagonal transition (WFA) [[file:imgs/readme/3_diagonal_transition.gif]] | DT + Divide & Conquer (BiWFA) [[file:imgs/readme/4_dt-divide-and-conquer.gif]] |
| A*PA (GCSH+DT) [[file:imgs/readme/5_astarpa.gif]]                        |                                                                            |

* Paper
  *Ragnar Groot Koerkamp*, *Pesho Ivanov*.
  "Exact pairwise alignment using A* with seed heuristic and match pruning".
  bioRxiv (2022). DOI [[https://doi.org/10.1101/2022.09.19.508631][2022.09.19.508631]]

- Figures ::
  Paper figures are generated using the examples at [[file:pa-bin/examples/paper-figures/][pa-bin/examples/paper-figures]].

- Evals ::
  Benchmarking code, evals, and datasets have moved to [[https://github.com/pairwise-alignment/pa-bench][pa-bench]].
  Results can be found in [[https://github.com/pairwise-alignment/pa-bench/blob/main/evals/astarpa/evals.ipynb][this notebook]] and reproduced using [[https://github.com/pairwise-alignment/pa-bench/blob/main/evals/astarpa/makefile][this makefile]].
  Dataset downloads are in [[https://github.com/pairwise-alignment/pa-bench/releases/tag/datasets][this release]].

- Tests ::
  Code is tested for correctness in various tests ([[file:astarpa/src/tests/][astarpa/src/tests]])
  against ~triple-accel~. The benchmark tool [[https://github.com/pairwise-alignment/pa-bench][pa-bench]] also checks correctness automatically.

- Benchmarks ::
  The code is benchmarked on GitHub Actions CI. Performance history of
  benchmarks is [[https://ragnargrootkoerkamp.github.io/astar-pairwise-aligner/dev/bench/][here]].

* Crate structure

Code is spread out over multiple crates.
From low to high:
- ~pa-types~: Basic types such as ~Seq~, ~Pos~, ~Cigar~, and ~Cost~, hosted in
  the ~pairwise-alignment~ org.
- ~pa-affine-types~: Types for affine edit graphs such as
   ~State = (Pos, Layer)~, ~AffineCigar~, and ~CostModel~. Not used by A*PA, but other
  algorithms and the visualizer support it.
- ~pa-heuristic~: Code for
  - finding matches
  - computing contours (fast and bruteforce)
  - heuristics themselves
  - wrapper/bruteforce heuristics for debugging
- ~pa-vis-types~: Trait definition of the visualizer callbacks, and the empty ~NoVis~ visualizer.
- ~astarpa~: Main A*PA API entrypoint containing the ~astar~ and ~astar_dt~
  functions, the ~bucket_queue~ data structure, and the ~astarpa(a,b)~ entrypoint.
- ~astarpa-c~: C-bindings for ~astarpa~
- ~pa-vis~: The visualizer. Contains a ~Canvas~ trait implemented for the
  ~SDL2Canvas~. The ~sdl2~ feature is optional.
- ~pa-generate~: Library and binary to generate different types of random sequences.
- ~pa-bin~: Main command line interface to A*PA. Allows for input from file,
  generated input, visualizing, and customization of the A*PA parameters.
- ~pa-bitpacking~: Implementation of Myers' bitpacking algorithms and SIMD extensions.
- ~astarpa2~: Work-in-progress code for the next version, A*PA2.
- ~pa-base-algos~: Re-implementations of Needleman-Wunsch/Edlib and
  Diagonal-transition/WFA/BiWFA for visualizations. Also playground for new A*PA
  variants such as [[https://curiouscoding.nl/posts/local-doubling/][local doubling]].
- ~astarpa-next~: Some code for other new ideas such as [[https://curiouscoding.nl/posts/speeding-up-astar/][pre-pruning]].
- ~pa-web~: web-interface to A*PA by compiling to webassembly. Implements the
  ~Canvas~ trait for ~HTMLCanvas~. (Not maintained.)

#+begin_src shell :results file :file imgs/readme/depgraph.svg :exports results
cargo depgraph --dedup-transitive-deps \
    --include pa-generate,pa-bin,pa-vis,astarpa,pa-types,pa-affine-types,sdl2,pa-base-algos,pa-heuristic,pa-vis-types,astarpa-c,pa-bitpacking,astarpa2,astarpa-next \
    | dot -T svg
#+end_src

#+RESULTS:
[[file:imgs/readme/depgraph.svg]]



* License
MPL-2.0
